components/user-mgt/org.wso2.carbon.identity.user.pre.update.password.action/src/main/java/org/wso2/carbon/identity/user/pre/update/password/action/execution/PreUpdatePasswordActionRequestBuilder.java

```diff
@@ -0,0 +1,219 @@
+
+        return actionRequestBuilder.build();
+    }
+
+    private void resolveUserActionContext(Map<String, Object> eventContext)
+            throws ActionExecutionRequestBuilderException {
+
+        Object action = eventContext.get(PreUpdatePasswordActionConstants.USER_ACTION_CONTEXT);
+        if (!(action instanceof UserActionContext)) {
+            throw new ActionExecutionRequestBuilderException("User Action Context cannot be null.");
+        }
+        userActionContext = (UserActionContext) action;
+    }
+
+    private void resolveAction(Map<String, Object> eventContext) throws ActionExecutionRequestBuilderException {
+
+        Object action = eventContext.get("action");
+        if (!(action instanceof PreUpdatePasswordAction)) {
+            throw new ActionExecutionRequestBuilderException("Pre Update Password Action cannot be null.");
+
+        }
+        preUpdatePasswordAction = (PreUpdatePasswordAction) action;
+    }
+
+    private Event getEvent() throws ActionExecutionRequestBuilderException {
+
+        return new PreUpdatePasswordEvent.Builder()
+                .initiator(getInitiator())
+                .action(getAction())
+                .tenant(getTenant())
+                .user(getUser())
+                .userStore(new UserStore(userActionContext.getUserStoreDomain()))
+                .build();
+    }
+
+    private PreUpdatePasswordEvent.FlowInitiator getInitiator() throws ActionExecutionRequestBuilderException {
+
+        Flow flow = IdentityContext.getThreadLocalIdentityContext().getFlow();
+        if (flow == null) {
+            throw new ActionExecutionRequestBuilderException("Flow is null in the Identity Context.");
+        }
+
+        switch(flow.getInitiatingPersona()) {
+            case ADMIN:
+                return PreUpdatePasswordEvent.FlowInitiator.ADMIN;
+            case APPLICATION:
+                return PreUpdatePasswordEvent.FlowInitiator.APPLICATION;
+            case USER:
+                return PreUpdatePasswordEvent.FlowInitiator.USER;
+            default:
+                break;
+        }
+        throw new ActionExecutionRequestBuilderException("Invalid initiator flow.");
+    }
+
+    private PreUpdatePasswordEvent.Action getAction() throws ActionExecutionRequestBuilderException {
+
+        Flow flow = IdentityContext.getThreadLocalIdentityContext().getFlow();
+        if (flow == null) {
+            throw new ActionExecutionRequestBuilderException("Flow is null in the Identity Context.");
+        }
+
+        switch (flow.getName()) {
+            case PASSWORD_UPDATE:
+                return PreUpdatePasswordEvent.Action.UPDATE;
+            case PASSWORD_RESET:
+                return PreUpdatePasswordEvent.Action.RESET;
+            case USER_REGISTRATION_INVITE_WITH_PASSWORD:
+                if (flow.getInitiatingPersona().equals(Flow.InitiatingPersona.ADMIN)) {
+                    return PreUpdatePasswordEvent.Action.INVITE;
+                }
```

Comment:
Here initiating persona can be an admin or app. Do we need to check this ?

Reply:
But in the returning flow (where the user instracts and change the password), we cannot identify whether the invite is done by an ADMIN or an APPLICATION. Hence removed the if check and considered ADMIN as the initiated persona